designs/plugins.md

```diff
@@ -0,0 +1,255 @@
+ - `getPluginsByType` - get all plugins of a given type
+
+
+#### Plugin List format
+
+This follows the same basic structure as the node list format:
+
+```
+[
+    {
+        "id": "test-plugin/test",
+        "name": "test",
+        "enabled": true,
+        "local": true,
+        "module": "test-plugin",
+        "plugins": [
+            {
+                "type": "foo",
+                "id": "my-test-plugin",
+                "module": "test-plugin"
+            }
+        ],
+        "editor": true,
+        "runtime": true,
+        "version": "1.0.0"
+    }
+]
+```
+
+Each entry in the array represents a separate set of plugins - ie a separate entry in
+the `package.json`. The properties are:
+
+ - `id` - the module/set
+ - `name` - the set name
+ - `enabled` - whether the plugin is enabled (*TODO* this is a hangover from nodes. Not sure if we'll
+ support dynamic enable/disable of nodes)
+ - `local` - whether the module is installed in `~/.node-red` or not - only local modules can be removed. (*TODO* another hangover
+     from nodes)
+ - `module` - the module name
+ - `plugins` - an array of the plugins provided by this set
+ - `editor` - boolean flag to say if this is an editor plugin (ie, provides an html file)
+ - `runtime` - boolean flag to say if this is a runtime plugin (ie, provides a js file)
+ - `version` - npm module version
+
+### Editor APIs
+
+`RED.plugins` will be a new module in the Editor. It will expose the functions:
+
+ - `RED.plugins.registerPlugin(id,definition)` - called by plugins to register themselves
+ - `RED.plugins.getPlugins(id)` - get a plugin by identifier
+ - `RED.plugins.getPluginsByType(type)` - get all plugins of a particular type (using the optional `type` field in the plugin definition)
+
+TODO: there may be other APIs to expose
+
+### Editor customisations
+
+For plugins that add to the UI, we need to identify what UI elements they may want
+to extend and have a plan in place to have proper APIs in place for them.
+
+ - Adding a custom sidebar - already available
+ - Adding custom menu items - the `RED.menu` api needs overhauling to allow easier customisation of the main menu
+ - Adding library types - this will be covered in a separate design note on the improved library api
+ - Adding custom widgets to the workspace footer - work in progress, need to formalise the existing API
+
+
+### Example plugin - nrlint
+
+*I am using nrlint as an example to demonstrate how it could work. This is not meant to be the formal design for nrlint's plugin system. All of the names/types/ids I've used here are placeholders*
+
+ - The Flow Linter tool will register a plugin called `nrlint`.
+ - It's `onadd` function will do the following:
+     - add a custom sidebar to the editor and all the logic to do its work.
+     - it will listen for the `registry:plugin-added` event so it knows when other plugins are added.
+       When that is called, it will check if the new plugin has a `type` of `nrlint-rule`.
```

Comment:
Can we assume specific order of loading plugin?   For example, plugins are loaded as alphabetical order, we can safely assume `nrlint-plugin-foobar`(type: nrlint-rule) will be loaded after `nrlint` has been loaded.

Reply:
I am trying to design it so the order of loading plugins does not matter - they can be loaded in any order.

If we support dynamically installing plugins (maybe not in the first version, but later on), then a user could install `nrlint-plugin-foobar` first and then discover they should also install `nrlint`. They would expect that to work.

If a plugin is able to have plugins of its own, then it needs to be written to support this out of order loading.

I describe how that could work later on in the design:

 - when a plugin is loaded its `onadd` function is called. In the case of `nrlint`, it can do two things:
    1. register a listener on the `registry:plugin-added` event - so it will be notified when new plugins are loaded
    2. use `getPluginsByType("nrlint-rule")` to discover the plugins that have already been loaded.

A plugin of `nrlint-rule` could then look like this: (Note, I haven't looked at the API you have created between nrlint and its rules... but I hope this shows what I'm thinking of...)
```
RED.plugins.registerPlugin("nrlint-plugin-foobar", {
   type: "nrlint-rule",
   doLinting: function(flow) { ... }
});
```

That means `nrlint`, when it discovers a new rule plugin has been loaded, can set `RED.plugins.getPlugin(id)` to get ahold of the plugin, and then call its `doLinting` function when it needs to.

Does that make sense?

The API a rule exposes needs designing (again, I haven't looked at what you have in nrlint today yet). For example, a rule may have its own settings the user could want to customise in the UI. The plugin definition could include meta data about the settings so the nrlint UI can generate the right dialog for it.... but this is all discussion for nrlint, not the general plugin system. 🙂